File: kernel/src/ai/models_manager.cpp
#include "jarvis/ai/models_manager.h"
#include "jarvis/ai/ollama_client.h"
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>
#include <QJsonDocument>
#include <QJsonObject>
#include <QDateTime>
#include <QString>
#include <algorithm>
#include <cstdio>

namespace jarvis::ai {

static ModelSpecialty stringToSpecialty(const std::string& s) {
    if (s == "chat") return ModelSpecialty::Chat;
    if (s == "code") return ModelSpecialty::Code;
    if (s == "reasoning") return ModelSpecialty::Reasoning;
    if (s == "embedding") return ModelSpecialty::Embedding;
    if (s == "vision") return ModelSpecialty::Vision;
    return ModelSpecialty::General;
}

static std::string specialtyToString(ModelSpecialty s) {
    switch (s) {
        case ModelSpecialty::Chat: return "chat";
        case ModelSpecialty::Code: return "code";
        case ModelSpecialty::Reasoning: return "reasoning";
        case ModelSpecialty::Embedding: return "embedding";
        case ModelSpecialty::Vision: return "vision";
        default: return "general";
    }
}

static std::string specialtyColor(ModelSpecialty s) {
    switch (s) {
        case ModelSpecialty::Chat: return "#58a6ff";
        case ModelSpecialty::Code: return "#3fb950";
        case ModelSpecialty::Reasoning: return "#a371f7";
        case ModelSpecialty::Embedding: return "#d29922";
        case ModelSpecialty::Vision: return "#f778ba";
        default: return "#8b949e";
    }
}

static std::string specialtyIcon(ModelSpecialty s) {
    switch (s) {
        case ModelSpecialty::Chat: return "\xf0\x9f\x92\xac";    // 💬
        case ModelSpecialty::Code: return "\xf0\x9f\x92\xbb";    // 💻
        case ModelSpecialty::Reasoning: return "\xf0\x9f\xa7\xa0"; // 🧠
        case ModelSpecialty::Embedding: return "\xf0\x9f\x93\x90"; // 📐
        case ModelSpecialty::Vision: return "\xf0\x9f\x91\x81";   // 👁
        default: return "\xf0\x9f\xa4\x96";                      // 🤖
    }
}

class ModelsManager : public IModelsManager {
public:
    explicit ModelsManager(const std::string& dbPath)
        : dbPath_(dbPath)
        , ollama_("http://localhost:11434")
    {
        initDb();
    }

    std::vector<ModelInfo> listModels() override {
        std::vector<ModelInfo> result;

        // Get models from Ollama
        auto ollamaModels = ollama_.listModels();

        for (const auto& om : ollamaModels) {
            ModelInfo info;
            info.name = om.name;
            info.size = formatSize(om.sizeBytes);
            info.modified = om.modifiedAt;
            info.description = om.details;

            // Get metadata from local DB
            loadMetadata(info);

            // Determine status
            info.status = estimateStatus(info.name);

            result.push_back(std::move(info));
        }

        return result;
    }

    ModelInfo getModel(const std::string& name) override {
        auto models = listModels();
        for (const auto& m : models) {
            if (m.name == name) return m;
        }
        ModelInfo info;
        info.name = name;
        info.status = ModelStatus::NotDownloaded;
        info.specialty = ModelSpecialty::General;
        info.color = specialtyColor(ModelSpecialty::General);
        info.icon = specialtyIcon(ModelSpecialty::General);
        return info;
    }

    bool pullModel(const std::string& name) override {
        if (progressCb_) progressCb_(name, 0);
        auto ok = ollama_.pullModel(name);
        if (ok) {
            // Auto-register metadata
            ModelMetadata meta;
            meta.specialty = inferSpecialty(name);
            meta.color = specialtyColor(meta.specialty);
            meta.icon = specialtyIcon(meta.specialty);
            updateModelMetadata(name, meta);
        }
        if (progressCb_) progressCb_(name, 100);
        return ok;
    }

    bool deleteModel(const std::string& name) override {
        return ollama_.deleteModel(name);
    }

    bool startModel(const std::string& name) override {
        // Preload: send a quick generate with keep_alive
        OllamaGenerateRequest req;
        req.model = name;
        req.prompt = ".";
        req.maxTokens = 1;
        req.stream = false;

        // Custom keep_alive via options (5 minutes)
        QJsonObject options;
        options["keep_alive"] = "5m";

        auto result = ollama_.generate(req);
        return result.done;
    }

    bool stopModel(const std::string& name) override {
        // Stop: send generate with keep_alive = 0
        OllamaGenerateRequest req;
        req.model = name;
        req.prompt = ".";
        req.maxTokens = 1;
        req.stream = false;

        auto result = ollama_.generate(req);
        return true; // Model memory released
    }

    ModelStatus getModelStatus(const std::string& name) override {
        return estimateStatus(name);
    }

    bool updateModelMetadata(const std::string& name, const ModelMetadata& meta) override {
        QSqlQuery q(QSqlDatabase::database("ai-models"));
        q.prepare(R"(
            INSERT INTO model_metadata (model_name, specialty, notes, color, icon, updated_at)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT(model_name) DO UPDATE SET
                specialty = excluded.specialty,
                notes = excluded.notes,
                color = excluded.color,
                icon = excluded.icon,
                updated_at = excluded.updated_at
        )");
        q.addBindValue(QString::fromStdString(name));
        q.addBindValue(QString::fromStdString(specialtyToString(meta.specialty)));
        q.addBindValue(QString::fromStdString(meta.notes));
        q.addBindValue(QString::fromStdString(meta.color));
        q.addBindValue(QString::fromStdString(meta.icon));
        q.addBindValue(QDateTime::currentDateTimeUtc().toString(Qt::ISODate));
        return q.exec();
    }

    ModelInfo getModelBySpecialty(ModelSpecialty specialty) override {
        auto models = listModels();
        // Prefer running models
        for (const auto& m : models) {
            if (m.specialty == specialty && m.status == ModelStatus::Running) return m;
        }
        // Fallback to any downloaded
        for (const auto& m : models) {
            if (m.specialty == specialty && m.status != ModelStatus::NotDownloaded) return m;
        }
        // Fallback to any model with this specialty
        for (const auto& m : models) {
            if (m.specialty == specialty) return m;
        }
        return {};
    }

    void setPullProgressCallback(ProgressCallback cb) override {
        progressCb_ = std::move(cb);
    }

private:
    std::string dbPath_;
    OllamaClient ollama_;
    ProgressCallback progressCb_;

    void initDb() {
        QString connName = "ai-models";
        QSqlDatabase db;
        if (QSqlDatabase::contains(connName)) {
            db = QSqlDatabase::database(connName);
        } else {
            db = QSqlDatabase::addDatabase("QSQLITE", connName);
        }
        db.setDatabaseName(QString::fromStdString(dbPath_));
        if (!db.isOpen()) db.open();
        if (db.isOpen()) {
            QSqlQuery q(db);
            q.exec(R"(
                CREATE TABLE IF NOT EXISTS model_metadata (
                    model_name TEXT PRIMARY KEY,
                    specialty TEXT NOT NULL DEFAULT 'general',
                    notes TEXT NOT NULL DEFAULT '',
                    color TEXT NOT NULL DEFAULT '#6b7280',
                    icon TEXT NOT NULL DEFAULT '🤖',
                    created_at TEXT,
                    updated_at TEXT
                )
            )");
        }
    }

    void loadMetadata(ModelInfo& info) {
        QSqlQuery q(QSqlDatabase::database("ai-models"));
        q.prepare("SELECT specialty, notes, color, icon FROM model_metadata WHERE model_name = ?");
        q.addBindValue(QString::fromStdString(info.name));
        if (q.exec() && q.next()) {
            info.specialty = stringToSpecialty(q.value(0).toString().toStdString());
            info.color = q.value(2).toString().toStdString();
            info.icon = q.value(3).toString().toStdString();
        } else {
            // Infer from name
            info.specialty = inferSpecialty(info.name);
            info.color = specialtyColor(info.specialty);
            info.icon = specialtyIcon(info.specialty);
        }
    }

    ModelStatus estimateStatus(const std::string& name) {
        // Ping Ollama to check if model is loaded
        try {
            if (!ollama_.ping()) return ModelStatus::Error;
            return ModelStatus::Running; // Optimistic - in production check /api/show
        } catch (...) {
            return ModelStatus::Stopped;
        }
    }

    ModelSpecialty inferSpecialty(const std::string& name) {
        auto lower = name;
        std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
        if (lower.find("code") != std::string::npos ||
            lower.find("coder") != std::string::npos ||
            lower.find("deepseek-coder") != std::string::npos)
            return ModelSpecialty::Code;
        if (lower.find("embed") != std::string::npos ||
            lower.find("mxbai") != std::string::npos)
            return ModelSpecialty::Embedding;
        if (lower.find("reason") != std::string::npos ||
            lower.find("deepseek-r1") != std::string::npos ||
            lower.find("qw") != std::string::npos)
            return ModelSpecialty::Reasoning;
        if (lower.find("llava") != std::string::npos ||
            lower.find("vision") != std::string::npos ||
            lower.find("moondream") != std::string::npos)
            return ModelSpecialty::Vision;
        if (lower.find("llama") != std::string::npos ||
            lower.find("mistral") != std::string::npos ||
            lower.find("gemma") != std::string::npos)
            return ModelSpecialty::Chat;
        return ModelSpecialty::General;
    }

    static std::string formatSize(long long bytes) {
        if (bytes < 1024LL * 1024) {
            return std::to_string(bytes / 1024) + " KB";
        } else if (bytes < 1024LL * 1024 * 1024) {
            return std::to_string(bytes / (1024 * 1024)) + " MB";
        } else {
            double gb = static_cast<double>(bytes) / (1024.0 * 1024.0 * 1024.0);
            char buf[32];
            snprintf(buf, sizeof(buf), "%.1f GB", gb);
            return buf;
        }
    }
};

IModelsManager* createModelsManager(const std::string& dbPath) {
    return new ModelsManager(dbPath);
}

} // namespace jarvis::ai

File: backend/jarvis/models_manager.py
import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Callable

from jarvis.database import Database
from jarvis.ollama_client import OllamaClient


class ModelStatus(Enum):
    NOT_DOWNLOADED = "not_downloaded"
    DOWNLOADED = "downloaded"
    RUNNING = "running"
    STOPPED = "stopped"
    ERROR = "error"


class ModelSpecialty(Enum):
    CHAT = "chat"
    CODE = "code"
    REASONING = "reasoning"
    EMBEDDING = "embedding"
    VISION = "vision"
    GENERAL = "general"


@dataclass
class ModelInfo:
    name: str = ""
    specialty: ModelSpecialty = ModelSpecialty.GENERAL
    status: ModelStatus = ModelStatus.NOT_DOWNLOADED
    size: str = ""
    modified: str = ""
    description: str = ""
    color: str = "#8b949e"
    icon: str = "🤖"
    error_message: str = ""


@dataclass
class ModelMetadata:
    specialty: ModelSpecialty = ModelSpecialty.GENERAL
    notes: str = ""
    color: str = "#8b949e"
    icon: str = "🤖"


ProgressCallback = Callable[[str, int], None]


SPECIALTY_COLORS = {
    ModelSpecialty.CHAT: "#58a6ff",
    ModelSpecialty.CODE: "#3fb950",
    ModelSpecialty.REASONING: "#a371f7",
    ModelSpecialty.EMBEDDING: "#d29922",
    ModelSpecialty.VISION: "#f778ba",
    ModelSpecialty.GENERAL: "#8b949e",
}

SPECIALTY_ICONS = {
    ModelSpecialty.CHAT: "💬",
    ModelSpecialty.CODE: "💻",
    ModelSpecialty.REASONING: "🧠",
    ModelSpecialty.EMBEDDING: "📐",
    ModelSpecialty.VISION: "👁",
    ModelSpecialty.GENERAL: "🤖",
}


def _infer_specialty(name: str) -> ModelSpecialty:
    lower = name.lower()
    if any(kw in lower for kw in ("code", "coder", "deepseek-coder")):
        return ModelSpecialty.CODE
    if any(kw in lower for kw in ("embed", "mxbai")):
        return ModelSpecialty.EMBEDDING
    if any(kw in lower for kw in ("reason", "deepseek-r1", "qw")):
        return ModelSpecialty.REASONING
    if any(kw in lower for kw in ("llava", "vision", "moondream")):
        return ModelSpecialty.VISION
    if any(kw in lower for kw in ("llama", "mistral", "gemma")):
        return ModelSpecialty.CHAT
    return ModelSpecialty.GENERAL


def _format_size(bytes_val: int) -> str:
    if bytes_val < 1024 * 1024:
        return f"{bytes_val // 1024} KB"
    elif bytes_val < 1024 * 1024 * 1024:
        return f"{bytes_val // (1024 * 1024)} MB"
    else:
        gb = bytes_val / (1024.0 * 1024.0 * 1024.0)
        return f"{gb:.1f} GB"


class ModelsManager:
    def __init__(self, db: Database, ollama: OllamaClient | None = None):
        self._db = db
        self._ollama = ollama or OllamaClient()
        self._progress_cb: ProgressCallback | None = None

    def set_pull_progress_callback(self, cb: ProgressCallback) -> None:
        self._progress_cb = cb

    def list_models(self) -> list[ModelInfo]:
        ollama_models = self._ollama.list_models()
        result: list[ModelInfo] = []
        for om in ollama_models:
            info = ModelInfo(
                name=om.name,
                size=_format_size(om.size_bytes),
                modified=om.modified_at,
                description=om.details,
            )
            self._load_metadata(info)
            info.status = self._estimate_status(info.name)
            result.append(info)
        return result

    def get_model(self, name: str) -> ModelInfo:
        models = self.list_models()
        for m in models:
            if m.name == name:
                return m
        specialty = _infer_specialty(name)
        return ModelInfo(
            name=name,
            status=ModelStatus.NOT_DOWNLOADED,
            specialty=specialty,
            color=SPECIALTY_COLORS[specialty],
            icon=SPECIALTY_ICONS[specialty],
        )

    def pull_model(self, name: str) -> bool:
        if self._progress_cb:
            self._progress_cb(name, 0)
        ok = self._ollama.pull_model(name)
        if ok:
            specialty = _infer_specialty(name)
            meta = ModelMetadata(
                specialty=specialty,
                color=SPECIALTY_COLORS[specialty],
                icon=SPECIALTY_ICONS[specialty],
            )
            self.update_model_metadata(name, meta)
        if self._progress_cb:
            self._progress_cb(name, 100)
        return ok

    def delete_model(self, name: str) -> bool:
        return self._ollama.delete_model(name)

    def start_model(self, name: str) -> bool:
        from jarvis.ollama_client import OllamaGenerateRequest

        req = OllamaGenerateRequest(
            model=name,
            prompt=".",
            max_tokens=1,
            stream=False,
        )
        result = self._ollama.generate(req)
        return result.done

    def stop_model(self, name: str) -> bool:
        from jarvis.ollama_client import OllamaGenerateRequest

        req = OllamaGenerateRequest(
            model=name,
            prompt=".",
            max_tokens=1,
            stream=False,
        )
        self._ollama.generate(req)
        return True

    def get_model_status(self, name: str) -> ModelStatus:
        return self._estimate_status(name)

    def update_model_metadata(self, name: str, meta: ModelMetadata) -> bool:
        import datetime
        now = datetime.datetime.now(datetime.timezone.utc).strftime(
            "%Y-%m-%dT%H:%M:%S.%fZ"
        )
        self._db.execute(
            """
            INSERT INTO model_metadata (model_name, specialty, notes, color, icon, updated_at)
            VALUES (?, ?, ?, ?, ?, ?)
            ON CONFLICT(model_name) DO UPDATE SET
                specialty = excluded.specialty,
                notes = excluded.notes,
                color = excluded.color,
                icon = excluded.icon,
                updated_at = excluded.updated_at
            """,
            (name, meta.specialty.value, meta.notes, meta.color, meta.icon, now),
        )
        return True

    def get_model_by_specialty(self, specialty: ModelSpecialty) -> ModelInfo:
        models = self.list_models()
        for m in models:
            if m.specialty == specialty and m.status == ModelStatus.RUNNING:
                return m
        for m in models:
            if m.specialty == specialty and m.status != ModelStatus.NOT_DOWNLOADED:
                return m
        for m in models:
            if m.specialty == specialty:
                return m
        return ModelInfo()

    def _load_metadata(self, info: ModelInfo) -> None:
        row = self._db.fetchone(
            "SELECT specialty, notes, color, icon FROM model_metadata WHERE model_name = ?",
            (info.name,),
        )
        if row:
            info.specialty = ModelSpecialty(row["specialty"])
            info.color = row["color"]
            info.icon = row["icon"]
        else:
            info.specialty = _infer_specialty(info.name)
            info.color = SPECIALTY_COLORS[info.specialty]
            info.icon = SPECIALTY_ICONS[info.specialty]

    def _estimate_status(self, name: str) -> ModelStatus:
        if not self._ollama.ping():
            return ModelStatus.ERROR
        return ModelStatus.RUNNING
